### PR TITLE
Recursive Downloads + Refactor

### PR DESCRIPTION
**Pull request by morsecodist**
This PR fixes recursive downloads as well as refactors the code and improves mmap. A summary of the changes:

- Recursive uploads, downloads, and local file copying (why not?)
- Simplifies mmap by leveraging the `copy` function
- Brings mmap behaviour in line with file's `ReadAt` and `Read` by using `io.EOF`
- Create an interface for paths so s3 path and file path operations can be performed interchangeably
- Replace `mains` with a unified path logic based on the `Path` interface
- Created separate options structs for various methods to decouple their options
- Refactored checksum getting/setting in s3 to use common methods
- Removed instances of panic for more user-friendly errors
- Added specific error messages for many cases

### CODE REVIEW

File: s3checksum/checksum.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	"strconv"
+
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+)
+
+// Crc32cChecksumMetadataName is the name of the metadata field to store the crc32c checksum
+// 	This capitalization is critical to the implementation please do not change it
+//	if you write metadata with different capitalization s3 will fuse it with
+//  the existing value of the same name instead of overwriting.
```
Comment (jshoe) [inline s3checksum/checksum.go]: 😅

File: s3utils/copy.go [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+	isDestDir, err := dest.IsDir()
+	if destExists && err != nil {
+		return []CopyJob{}, err
+	}
+
+	if !isDestDir && isSrcDir {
+		if !destExists {
+			// If the destination doesn't exist and the source is a directory
+			//   create a local directory. This brings local behavior in line
+			//   with s3, where it is possible to upload to a non-existant folder
```
Comment (jshoe) [inline s3utils/copy.go]: nit: existent

File: s3utils/copy.go [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+	// Only compute checksum if it is necessary
+	if c.Options.Checksum {
+		checksumOptions := checksum.ParallelChecksumOptions{
+			Concurrency: c.Options.Concurrency,
+			PartSize:    c.Options.PartSize,
+			UseMmap:     c.Options.Mmap,
+		}
+		crc32cChecksum, err := checksum.ParallelCRC32CChecksum(src, checksumOptions)
+		if err != nil {
+			return errors.New("Error computing crc32c checksum of source file")
```
Comment (jshoe) [inline s3utils/copy.go]: would the error from `ParallelCRC32CChecksum` be helpful to write out here?

File: s3utils/copy.go [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+		}
+		uploadInput = s3checksum.SetCRC32CChecksum(uploadInput, crc32cChecksum)
+	}
+
+	// TODO make an mmap API that is compatible with os.File to avoid this branching
```
Comment (jshoe) [inline s3utils/copy.go]: might be helpful to add these TODOs to the Github issues?

File: s3utils/copy.go [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+}
+
+// Copy executes a copy job
+func (c *Copier) Copy(copyJob CopyJob) error {
+	if copyJob.source.IsS3() && copyJob.destination.IsS3() {
+		return errors.New("Copying between s3 is not yet supported")
+	} else if !copyJob.source.IsS3() && copyJob.destination.IsS3() {
+		bucket, err := copyJob.destination.Bucket()
+		if err != nil {
+			return fmt.Errorf("path: %s was determined to be an s3 path but getting it's bucket encountered error: %s", copyJob.destination, err)
```
Comment (jshoe) [inline s3utils/copy.go]: nit: its

File: s3utils/localpath.go [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+func (p localPath) IsDir() (bool, error) {
+	stat, err := os.Stat(p.raw)
+	if err != nil {
+		return false, err
+	}
+
+	return stat.IsDir(), nil
+}
+
+// Exists Checks if a localPath is a directory
```
Comment (jshoe) [inline s3utils/localpath.go]: This comment for `Exists` is wrong, right?

File: s3utils/localpath.go [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+	return path.Base(p.raw)
+}
+
+// ToStringWithoutBucket returns a raw string path without the s3 bucket
+func (p localPath) ToStringWithoutBucket() string {
+	return p.raw
+}
+
+// WithoutPrefix returns a string representation of this path with the
+//   prefixPath (ignoring it's s3 bucket) removed from the begining. This is
```
Comment (jshoe) [inline s3utils/localpath.go]: nit: its

File: s3utils/s3path.go [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+	})
+
+	if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == "NotFound" {
+		return false, nil
+	}
+
+	return true, err
+}
+
+// IsS3 checks if a s3Path is an s3Path (it will never be)
```
Comment (jshoe) [inline s3utils/s3path.go]: These comments are mixed up "it will never be"

File: s3utils/path.go [hunk 9]
```diff
@@ -1,24 +1,47 @@
...
+}
+
+func bucketAndKeyToS3Path(bucket string, key string) string {
+	return fmt.Sprintf("s3://%s", path.Join(bucket, key))
 }
 
-// IsS3Path checks whether a string is an s3 path
-func IsS3Path(path string) bool {
+// isS3Path checks whether a string is an s3 path
+func isS3Path(path string) bool {
```
Comment (jshoe) [inline s3utils/path.go]: This capitalization looks like a bug b/c you're calling `s3utils.IsS3Path` elsewhere

File: s3utils/localpath.go (hunk 7 above)
Comment (jshoe) [reply s3utils/localpath.go]: `WithoutPrefix` seems a little ambiguous to me b/c you might expect prefix to just mean `s3://`. Maybe just `WithoutBucket` or `ObjectKey`?

File: s3utils/path.go (hunk 9 above)
Comment (morsecodist) [reply s3utils/path.go]: I didn't see any usages of `s3utils.IsS3Path`. The way this should be done is:

```go
path, _ := s3utils.NewPath(client, myPath)
path.IsS3()
```

`isS3Path` is a private function for internal use by `NewPath` only

File: s3utils/localpath.go (hunk 7 above)
Comment (morsecodist) [reply s3utils/localpath.go]: `WithoutPrefix` is definitely confusing it actually removes one path from another ignoring the scheme (`s3://`) and bucket. I will just remove the method completely, it is only used in one place so I will just put the implementation there. It is actually less code and less confusing that way imo.

Comment (morsecodist) [reply s3utils/localpath.go]: Also I will rename `ToStringWithoutBucket` to `WithoutBucket` it has a `string` return type so the name is redundant.

File: s3utils/path.go (hunk 9 above)
Comment (jshoe) [reply s3utils/path.go]: ah, I was looking at the deleted side 👍

**Comment by morsecodist** [comment]
@jshoe Small update added `--recursive` option to bring the api in line with `s3cp` also printed an options parsing error that I should have printed originally. I added some utility functions to help get the names right when printing error messages.

File: s3utils/localpath.go [hunk 10]
```diff
@@ -43,6 +43,16 @@ func (p localPath) IsLocal() bool {
 	return true
 }
 
+// DirOrFolder returns "directory" for localPath and "folder" for s3Path
+func (p localPath) DirOrFolder() string {
+	return "directory"
+}
+
+// FileOrObject returns "object" for localPath and "file" for s3path
```
Comment (jshoe) [inline s3utils/localpath.go]: this comment is reversed, right? file for localPath

File: s3utils/s3path.go [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+func (p s3Path) IsLocal() bool {
+	return false
+}
+
+// DirOrFolder returns "directory" for localPath and "folder" for s3Path
+func (p s3Path) DirOrFolder() string {
+	return "folder"
+}
+
+// FileOrObject returns "object" for localPath and "file" for s3path
```
Comment (jshoe) [inline s3utils/s3path.go]: this comment is reversed, right? file for localPath

**Comment by jshoe** [comment]
Still looks good w/ comment 👍

File: s3utils/localpath.go (hunk 10 above)
Comment (morsecodist) [reply s3utils/localpath.go]: :man_facepalming: second time I've done this thanks!

**Pull request merged**